### PR TITLE
ZKR-2589-Lookups-Unterpreted-Functions

### PR DESCRIPTION
**Pull request by FatemehHeidari**
This PR introduces the option to replace extensive lookup constraints with uninterpreted functions in circuits containing lookups. This method significantly accelerates the analysis process but comes with the caveat of potentially introducing false positives. To address this, a post-processing step is implemented to validate models found as satisfying for a fixed set of public inputs by verifying the values against the lookup table.

Key features and changes include:

An interactive option for users to opt into using uninterpreted functions for lookups, enhancing flexibility based on their performance and accuracy needs.
- The addition of the `decompose_lookups_as_uniterpreted()` function for extracting essential data for post-processing results. This includes:
    - Generating a vector of hashmaps for each lookup, mapping cells to their corresponding index in the fixed column.
- Introduction of `decompose_lookup_expression()` function to decompose lookup expressions, given the more simple structure of lookup expressions (selector * advice/instance = fixed) in supported versions of halo2.
- Implementation of the `lookup()` function to perform searches in lookup-dependent cells within fixed columns, further refining the analysis accuracy.

These changes have been thoroughly tested on circuits employing multiple lookups for all supported halo2 versions.

### CODE REVIEW

File: korrekt/src/circuit_analyzer/analyzer.rs [hunk 1]
```diff
@@ -776,6 +918,170 @@ impl<'b, F: AnalyzableField> Analyzer<F> {
...
+        feature = "use_zcash_halo2_proofs",
+        feature = "use_pse_halo2_proofs",
+        feature = "use_axiom_halo2_proofs",
+        feature = "use_pse_v1_halo2_proofs",
+    ))]
+    fn decompose_lookups_as_uniterpreted(
+        &mut self,
+        printer: &mut smt::Printer<File>,
+    ) -> Result<(), anyhow::Error> {
+        // As we are defining one uninterpreted function for all lookups, we only need to define it once.
```
Comment (jgorzny) [inline korrekt/src/circuit_analyzer/analyzer.rs:933]: I think this needs discussion -- not all lookups look at the same inputs, right? So I don't think this is the ideal approach. Later, if you try to include equality for these functions, they may actually be different lookups (with different ranges).

**Review by jgorzny — COMMENTED** [review]
I left a comment on the lookup design that I think should be reviewed or justified.

File: korrekt/src/circuit_analyzer/analyzer.rs [hunk 2]
```diff
@@ -633,6 +642,129 @@ impl<'b, F: AnalyzableField> Analyzer<F> {
...
+                );
+                let mut t = term.clone();
+                if cell_to_cycle_head.contains_key(&term) {
+                    t = cell_to_cycle_head[&term.clone()].to_string();
+                }
+                smt::write_var(printer, t.to_string());
+                (t.to_string(), NodeType::Advice, t.to_string())
+            }
+            Expression::Negated(poly) => {
+                todo!()
```
[the remark this answers is not in the document]
Comment (FatemehHeidari) [reply korrekt/src/circuit_analyzer/analyzer.rs]: I added an `Invalid` node type for expression types that we don't expect to see in a lookup expression, and removed these todos(). Added a TODO tag to add error handling for these invalid expressions.

File: korrekt/src/circuit_analyzer/analyzer.rs (hunk 1 above)
Comment (FatemehHeidari) [reply korrekt/src/circuit_analyzer/analyzer.rs]: Fixed this based on our discussion. Now, for each lookup we will have a separate uninterpreted function for each lookup.

File: korrekt/src/circuit_analyzer/analyzer.rs [hunk 3]
```diff
@@ -909,105 +1222,170 @@ impl<'b, F: AnalyzableField> Analyzer<F> {
...
+                same_str.push_str(&or_diff_assignments);
+                let and_all = smt::get_and(printer, same_str);
+                smt::write_assert_bool(printer, and_all, Operation::And);
+
+                // 4. find a model that satisfies these rules
+                let model_with_constraint =
+                    Self::solve_and_get_model(smt_file_path.clone(), &variables)
+                        .context("Failed to solve and get model!")?;
+
+                // If uaing uninterpreted function, we need to check if the model is valid by performing the lookup.
```
Comment (jgorzny) [inline korrekt/src/circuit_analyzer/analyzer.rs:1344]:
```suggestion
                // If using uninterpreted function, we need to check if the model is valid by performing the lookup.
```

File: korrekt/src/circuit_analyzer/analyzer.rs [hunk 4]
```diff
@@ -909,105 +1222,170 @@ impl<'b, F: AnalyzableField> Analyzer<F> {
...
+                    for r in &model_with_constraint.result {
+                        info!("{} : {}", r.1.name, r.1.value.element)
+                    }
+                    uc_lookup_dependency = false;
+                    for index in 0..self.lookup_mappings.len() {
+                        // Lookup search to make sure all values in the model are valid.
+                        let lookup_sucessful = self
+                            .lookup(&model_with_constraint, index)
+                            .context("Failed to perform lookup")?;
+                        // if the lookup is not successful, the model found is not valid and the under-constraiend flag is a false positive, still we can't say if the circuit is under-constrained or not.
```
Comment (jgorzny) [inline korrekt/src/circuit_analyzer/analyzer.rs:1356]:
```suggestion
                        // if the lookup is not successful, the model found is not valid and the under-constrained flag is a false positive, still we can't say if the circuit is under-constrained or not.
```

File: korrekt/src/circuit_analyzer/analyzer.rs [hunk 5]
```diff
@@ -776,6 +914,176 @@ impl<'b, F: AnalyzableField> Analyzer<F> {
...
+                                    region_begin,
+                                    region_end,
+                                    i32::try_from(row_num).ok().unwrap(),
+                                    &region.enabled_selectors,
+                                    &self.fixed,
+                                    &self.cell_to_cycle_head,
+                                );
+                                cons_str_vec.insert(node_str);
+                                if !var.is_empty() {
+                                    //TODO: check if we can optimize analyzer based on disabled selectores on lookups: ZKR-3207
```
Comment (jgorzny) [inline korrekt/src/circuit_analyzer/analyzer.rs:1036]: This would be great!

**Review by jgorzny — APPROVED** [review]
Two minor suggestions, too.

**Pull request merged**